### DISCUSSION TITLE
Can we override

### DISCUSSION THREAD

**Discussion by skuflow** [oqtane/oqtane.framework#2922] [Q&A]
is it feasible to override core functionality like the "remember me" function?  We have a client with a very particular way they want this to work.  It would be nice if we could override core methods and change functionality as needed.  Is this possible?

**Comment by W6HBR** [comment]
You can override anything you want.  Just download the source, make the changes, compile and deploy.  I personally have a few tweaks for my installs that I do this for.  I have a log of my changes and use Winmerge to do a code-compare each time I'm going to update to a newer version so that I can move my changes into the next version.  It's a little work, but not difficult at all.

Having said that, you can also pose your changes here and it may be well received and incorporated as a change to the base system.  I've had a few wants met this way.

**Reply by leigh-pointer** [reply]
@W6HBR 
When working with Oqtane, it's generally recommended not to modify the original source code directly. This is because if you make changes to the source, upgrading the framework later can become very complicated.

Oqtane is designed to be flexible and easy to customize. It follows a modular approach, which means you can extend its functionality without modifying the core code. Instead, you can create your own separate module based on the core framework and make modifications specific to your needs. You can add, remove, or customize features in this module.

For example, let's say you want to add or remove some functionality related to the user login process. Instead of modifying the original User Login module provided by Oqtane, you can create your own module specifically for handling user login. This way, the core framework remains intact and can be upgraded easily in the future if needed.

By following this approach, you can customize Oqtane to suit your requirements without getting tangled in upgrade complications.

**Comment by sbwalker** [comment]
@skuflow based on the fact that Oqtane is modular, you could create your own Login module and replace the default module on the Login page with your own (this is all managed by dynamic configuration which makes changes very simple). You could use the Login module code as a starting point (Oqtane.Client/Admin/Modules/Login/Index.razor) for your own module.  The benefit is that you do not need to modify the core framework - and you get the custom functionality you need. Any of the default administrative modules in Oqtane can be "overridden" using this approach (which is one of the main benefits of "modularity")

**Reply by vnetonline** [reply]
@sbwalker can you do a quick tutorial  on how we can use our own module and replace the existing ones without modifying the Oqtane framework ?

**Comment by sbwalker** [comment]
@vnetonline An Oqtane application is composed at run-time based on configuration defined in a database. By default every Oqtane site is created with a login page which contains a login module. So basically you simply need to change the configuration in the database so that the login page references your custom login module. There are many ways to do this depending on your requirements. 

1. Use the Admin UI. Navigate to the Login page, go into Edit mode, remove the Login module, add your own Login module. 
2. Modify the database. Find the record in the Module table which is related to the Login module. Override the ModuleDefinitionName to match your custom Login module name. 
3. Create a custom SiteTemplate. You can use the default SiteTemplates in the framework as an example. Include your custom Login module in the SiteTemplate. 
4. Use ISiteMigration to perform an automated update to the configuration of an existing site. 
5. Use the new PageTemplate concept in Oqtane 4.0 to update the configuration for the Login page to include your custom login module.

There are other methods as well but I would assume that one of the five methods above should meet your needs.
